### PR TITLE
(868) Assessment dashboard

### PR DESCRIPTION
**Pull request by pezholio**
This PR adds functionality for a user to view all their upcoming and approved assessments. There is also the ability to show asssessments that have had further information requested, but this is not yet baked into the API, so the tab for that information is currently blank.

I've tried to seperate concerns out in the PR a bit, with the clients dealing with the fetching of the data, the services marshalling the data into the correct format (in this case seperating the assessments by type) and the helper methods dealing with the presentation layer. This makes testing easier, and also helps make the functionality of each part of the application clearer. Before a bunch of presentation layer stuff was getting mixed in with the service layer (my fault!) and I think this made the service layer unweildy and difficult to use in other parts of the application. 

I've also broken down some of the bigger helper functions into smaller sub functions, which I then unit test, this makes testing more simple when we test the table row methods, so we don't have to spend ages setting up state to test every edge case is covered.

## Screenshots

Assess -- shows a list of assessments

Assess -- shows a list of assessments (1)

### CODE REVIEW

**Comment by pezholio** [comment]
[Assessment dashboard](https://trello.com/c/WBe1ppRb/868-assessment-dashboard)

File: server/utils/assessmentUtils.test.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  describe('daysSinceReceived', () => {
+    it('returns the difference in days since the assessment has been received', () => {
+      const assessment = assessmentFactory.createdXDaysAgo(10).build()
+
+      expect(daysSinceReceived(assessment)).toEqual(10)
+    })
+  })
+
+  describe('daysUntilDue', () => {
```
Comment (bryangaledxw) [inline server/utils/assessmentUtils.test.ts:41]: Looks like you've got the test for this in ahead of the function itself - function itself is in the next commit

File: server/utils/assessmentUtils.test.ts [hunk 2]
```diff
@@ -6,6 +6,9 @@ import {
   awaitingAssessmentTableRows,
   daysSinceInfoRequest,
   requestedFurtherInformationTableRows,
+  completedTableRows,
+  daysUntilDue,
+  formatDays,
```
Comment (bryangaledxw) [inline server/utils/assessmentUtils.test.ts:11]: I think have a look at the imports here, and the exports in `assessmentUtils.ts` - looks like they've gotten out of sync with when they're need and should have been included across the last couple of commits

File: server/utils/assessmentUtils.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        html: tierBadge(assessment.application.person.risks.tier.value.level),
+      },
+      {
+        text: formattedArrivalDate(assessment),
+      },
+      {
+        text: assessment.application.person.prisonName,
+      },
+      {
+        text: formatDays(daysUntilDue(assessment)),
```
Comment (bryangaledxw) [inline server/utils/assessmentUtils.ts]: Think this is being used before this function exists

File: server/utils/assessmentUtils.ts [hunk 4]
```diff
@@ -133,6 +133,16 @@ const formatDays = (days: number): string => {
   return `${days} Day${days > 1 ? 's' : ''}`
 }
 
+const formatDaysUntilDueWithWarning = (assessment: Assessment): string => {
+  const days = daysUntilDue(assessment)
+  if (days < DUE_DATE_APPROACHING_DAYS_WINDOW) {
+    return `<strong class="assessments--index__warning">${formatDays(
+      days,
+    )}<span class="govuk-visually-hidden">(Approaching due date)</span></strong>`
```
Comment (bryangaledxw) [inline server/utils/assessmentUtils.ts:141]: Do we want a space before the <span> tag, so it's not "4 Days(Approaching due date)"?

File: server/views/assessments/index.njk [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+              text: "Days until application due"
+            },
+            {
+              text: "Status"
+            }
+          ],
+          rows: AssessmentUtils.awaitingAssessmentTableRows(assessments.awaiting)
+        })
+      }}
+      {% endset -%}
```
Comment (bryangaledxw) [inline server/views/assessments/index.njk:47]: Because it's impossible to google - what's the difference between `{% endset %}` and `{% endset -%}`

File: integration_tests/tests/assess/listing.cy.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    assessments.push(assessmentsAwaiting)
+
+    // And two of those assessments are running close to the due date
+    const assessmentsCloseToDueDate = [
+      assessmentFactory.createdXDaysAgo(8).build({ decision: undefined }),
+      assessmentFactory.createdXDaysAgo(8).build({ decision: undefined }),
+    ]
+
+    assessments.push(assessmentsCloseToDueDate)
+    // And some completed assessments
```
Comment (bryangaledxw) [inline integration_tests/tests/assess/listing.cy.ts:31]: "And there are some completed assessments"?

**Review by bryangaledxw — APPROVED** [review]
Code looks good, think there's just some fixup to do with commit ordering around assessmentUtils.ts

File: server/data/assessmentClient.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import type { Assessment } from '@approved-premises/api'
```
[the remark this answers is not in the document]
Comment (bryangaledxw) [reply server/data/assessmentClient.ts:1]: n/m, spotted it later on

File: server/testutils/factories/assessment.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import type { Assessment } from '@approved-premises/api'
+
+import { DateFormats } from '../../utils/dateUtils'
+import applicationFactory from './application'
+
+class AssessmentFactory extends Factory<Assessment> {
+  createdXDaysAgo(days: number) {
+    const today = new Date()
+    return this.params({
+      createdAt: DateFormats.dateObjToIsoDate(new Date(today.getFullYear(), today.getMonth(), today.getDate() - days)),
```
[the remark this answers is not in the document]
Comment (pezholio) [reply server/testutils/factories/assessment.ts:12]: Javascript eh?

File: server/views/assessments/index.njk (hunk 5 above)
Comment (pezholio) [reply server/views/assessments/index.njk:47]: The trailing `-` stops it from including a linebreak in the generated HTML I think - I copied and pasted that bit from the GOV.UK docs 😄

File: server/services/assessmentService.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    const assessments = await client.all()
+
+    await Promise.all(
+      assessments.map(async assessment => {
+        const risks = await personClient.risks(assessment.application.person.crn)
+        const assessmentWithRisks = {
+          ...assessment,
+          application: { ...assessment.application, person: { ...assessment.application.person, risks } },
+        } as AssessmentWithRisks
+        return assessment.decision
```
[the remark this answers is not in the document]
Comment (pezholio) [reply server/services/assessmentService.ts:25]: Yeah, this gave me a `no-unused-expressions` eslint error, so I left it in https://eslint.org/docs/latest/rules/no-unused-expressions
